### ISSUE TITLE
Regent: import doesn't work with DCR

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#766]
We currently use semantic information on imported regions/index spaces/partitions to make sure we don't import them multiple times. These checks fail in a DCR context because one of the shards may arrive after the semantic information has already been attached on another node.

It occurs to me that we may be able to relax these checks for index spaces and partitions; only regions ought to really be unsafe if imported multiple times. I'm not sure what to do about regions though.

**Comment by magnatelee** [comment]
I actually hit this issue while I was working on the SC'19 paper (and I should have posted this issue then!), and I asked @lightsighter whether the operations on semantic information in shards can be replicated so that shards can be in sync with each other on those operations. I remember @lightsighter told me potential issues in supporting this, I don't remember what they were. @lightsighter for everyone's benefit, can you remind us of the issues?

**Comment by elliottslaughter** [comment]
At least one issue is that the first check is a check for *non*-existence. After that we attach the semantic info (which is replicated on all the nodes). So I'm not sure how to do this, unless we make the semantics of the query for semantic info to be to *not* check remote data if you're in a control replicated context, or something like that.

Am I missing something?

**Comment by magnatelee** [comment]
That's what I was proposing. Queries for semantic info will only use what is available in its replicated context and attach operations likewise will modify its local state. As long as all shards perform those operations deterministically in their program order, they will maintain a consistent view only with local accesses and mutations.

**Comment by magnatelee** [comment]
I guess one potential issue, which btw is somewhat orthogonal to the use case we discussed, is a race between child tasks and parent shards. Because queries from shards are neither deferred nor synchronized properly with child tasks,  it's not safe for the parent shard to query the semantic info attached by its child task. So using semantic information as an indicator for anything may not be a good idea in a distributed setting. That said, I'm not sure what else we can do other than changing the semantics of semantic information operations.

**Comment by lightsighter** [comment]
There are many issues with this approach, one of which is that you guys are kind of abusing what semantic information is for. 

> So using semantic information as an indicator for anything may not be a good idea in a distributed setting

That pretty much aligns with my viewpoint as well. 

Are imports global across tasks in a single process or are they something that is local to the lifetime of the task? If it's the later, you should probably be using task-local variables to track state within the task. 
https://gitlab.com/StanfordLegion/legion/-/blob/master/runtime/legion.h#L6328-6353
If it's the former, we'll need to come up with another plan.

**Comment by elliottslaughter** [comment]
As of master in commit fb9156045b8c438066a80d8bcdbca9a7f97c4d36, I have allowed multiple imports of index spaces and partitions (by just removing the tagging entirely). @magnatelee and I believe this should be safe, so the restrictions were unnecessary.

That still leaves regions, but I believe this is sufficient to unblock S3D for the time being.

**Comment by elliottslaughter** [comment]
Removing the S3D tag and bumping the milestone since this seems to be sufficient for S3D at the moment.

**Comment by lightsighter** [comment]
@elliottslaughter can you answer my question from before:

> Are imports global across tasks in a single process or are they something that is local to the lifetime of the task?

**Comment by elliottslaughter** [comment]
Currently it's a global property, but I think we can make it task-local. Basically, we need to make sure that any imported region does not alias with any region a task might currently have access to, which I think we can do with task-local tracking. (We will need to be careful to mark task arguments, which I think is not something we currently do, since the current tracking is global.)

**Comment by lightsighter** [comment]
Ok, it seems like locally scoping this to a task is the right answer in the long run anyway and the runtime I believe already provides the necessary facilities for doing this that do not require the use of semantic information.

**Comment by elliottslaughter** [comment]
Putting this on the backlog since I don't think anyone depends on this at the moment.

**Comment by elliottslaughter** [comment]
Is this still relevant?

**Comment by lightsighter** [comment]
Did Regent's import support move off of using semantic information attaches?

**Comment by elliottslaughter** [comment]
I think I followed through with this observation from my original comment:

> It occurs to me that we may be able to relax these checks for index spaces and partitions; only regions ought to really be unsafe if imported multiple times. I'm not sure what to do about regions though.

So, we do not tag index spaces or partitions. We do tag regions. I think people don't import regions in practice so no one is seeing this error, but I suppose it would pop up again if someone did try to do that.
